BIP 331: Ancestor Package Relay

glozow opened the pull request:
ML thread: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-May/020493.html

Supersedes #1324.

glozow commented:
Updated for number

ajtowns commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,373 @@
+'''No Package Information Round:''' Instead of having a package information round, just use the
+child's wtxid to refer to the package and always send the entire package together. This would make
+it very likely for honest nodes to redownload duplicate transactions.
+[[File:./bip-package-relay/no_package_info.png|900px]]
```
Need to update paths to images now that bip number is assigned

glozow replied:
Doh, thanks!

naumenkogs commented:
---------------

1. There are obvious typos and misalignments you're probably aware of, i assume we shouldn't merge them?

2.
> Package Information Only

This example seems confusing: it violates the following rule `A node MUST NOT send a "ancpkginfo" message that has not been requested by the recipient. Upon receipt of an unsolicited "ancpkginfo", a node may disconnect the sender.`
It will likely work if you expand the example to 3 transactions?

3. Unclear what to do when transactions are replaced in the middle of this process. Should that be covered? Probably will become clearer when you prototype.

4. Same about timestamping concerns i guess.

5. As a style comment, I'd suggest doing some compression of the text :)

6. `Is Package Erlay possible?`
6a) I actually made some attempt to handle parent/child within Erlay, do you think we should exercise the overlap between the two better? E.g., for those packages flowing through the network (not the IBD case, but fee-bump or something) Erlay might do the batching you need.
6b) One might also think about "what if we reconcile packages"? Probably bad idea, because the sets (packages) are too small? My point is that this should not be confused with what you describing.

7. MSG_ANCPKGINFO and getpkgtxns have some redundancy in the data... The latter could refer to the hash of the former + the indexes, for example? Not sure savings worth the complexity (how often this will be used? every lightning closure?) but maybe worth making it into q/a :)

glozow commented:
Thanks for the review @naumenkogs! yes, I think it's best to talk specifics with the actual implementation but I'll throw some quick answers for now:

> Unclear what to do when transactions are replaced in the middle of this process. Should that be covered? Probably will become clearer when you prototype.

One should treat this situation the same way we do for transactions that get replaced in between inv and getdata (I don't see any reason it should be different). Currently, we send "notfound". If in the future we decide to relay recently-replaced transactions, then package relay would include recently-replaced transactions as well. Though I agree this seems more like an implementation question than a spec question. In the spec, I would leave it as "the receiver may respond with notfound."

> I actually made some attempt to handle parent/child within Erlay, do you think we should exercise the overlap between the two better? E.g., for those packages flowing through the network (not the IBD case, but fee-bump or something) Erlay might do the batching you need.

Yes it's great that Erlay orders them correctly. Note below-fee-filter transactions to the reconciliation bucket so not every case is handled. I plan to have a rebase-on-top-of-Erlay version so we can exercise the two together.

> One might also think about "what if we reconcile packages"? Probably bad idea, because the sets (packages) are too small? My point is that this should not be confused with what you describing.

Mostly the "Is Package Erlay possible" section was to clarify that they aren't mutually exclusive.
Reconciling packages is I think something that may come up when we get to sender-initiated package relay. I'm thinking, after reconciliation, it may be useful to send a package inv instead of individual tx invs if there are fee-bumps... But for now, since packages must be requested by the receiver, we don't need to think about that yet.

> MSG_ANCPKGINFO and getpkgtxns have some redundancy in the data... The latter could refer to the hash of the former + the indexes, for example? 

Yeah that's an interesting idea, we could do a bitfield of 4 bytes (should cover any package since max count is 25). However this means you can only request getpkgtxns from somebody who sent an ancpkginfo (currently this is not a requirement). It also requires the sender to keep track of (allocate memory for, remember to expire) what they announced so they know what "the ith transaction from this package" means. Thoughts?

naumenkogs commented:
>Though I agree this seems more like an implementation question than a spec question. In the spec, I would leave it as "the receiver may respond with notfound."

The thing with current `notfound` is that it's atomic I guess — makes it easy to consider every announcement independent and our code treats it as such (in a poor way but there are also no expectations). With packages, I imagine a CPFP transaction gets replaced and dropped from the package due. A package recipient would then do what, drop the entire package, or consider it invalid?

A requestor will get this package which doesn't pass the fees, and drop it on the floor? Or should it re-request the package? At the moment, I'm not interested in the final answer — but interested in making sure we don't miss such questions.

> Reconciling packages is I think something that may come up when we get to sender-initiated package relay. I'm thinking, after reconciliation, it may be useful to send a package inv instead of individual tx invs if there are fee-bumps... But for now, since packages must be requested by the receiver, we don't need to think about that yet.

Not sure this anything to do with who initiates packages. Let's think about the current proposal. Imagine, upon the announcement of C, the receiver sends getdata(MSG_ANCPKGINFO, C), along with some compressed data saying that he already has 1 parent and 1 child of C (by just looking at the wtxid). Then it would save us announcing these 2 wtxids in the reverse direction.

>Yeah that's an interesting idea, we could do a bitfield of 4 bytes (should cover any package since max count is 25). However this means you can only request getpkgtxns from somebody who sent an ancpkginfo (currently this is not a requirement). It also requires the sender to keep track of (allocate memory for, remember to expire) what they announced so they know what "the ith transaction from this package" means. Thoughts?

I don't think memory allocation and stuff is a show-stopper, but it might be too much complexity if the gain is low. We should measure how much savings it allows... I think saving 28 byte per announced tx might be worthy. As for the "same peer", do you ever expect packages to be requested from a different peer from the one announced them?

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+ever need to use txid-based transaction relay.
+
+A single use case of txid-based relay remains: handling "orphan" transactions that spend
+output(s) from a transaction they are unaware of.  Orphans are common for new nodes that have just
```
```suggestion
output(s) from a transaction the receiving node is unaware of.  Orphans are common for new nodes that have just
```

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+(prevouts specify txid, not wtxid). These parents may end up being orphans as well, if they also
+spend unconfirmed inputs that the node is unaware of. This method of handling orphans is problematic
+for two reasons: it requires nodes to allocate memory for unvalidated data received on the p2p
+network and txid-based relay between two wtxid-relay peers.
```
```suggestion
network and it relies on txid-based relay between two wtxid-relay peers.
```

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+A transaction's '''descendants''' include, recursively, its children, the children of its children,
+etc. A transaction's parent is its ancestor, but an ancestor is not necessarily a parent.
+
+A '''package''' is an ordered list of transactions, representable by a connected Directed Acyclic
```
"an ordered list". Is there a well-defined ordering? What is it?

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+
+* Package {A, B} where A pays 0 satoshis and B pays 8000 satoshis in fees.
+* Package {C, D} where C pays 0 satoshis and D pays 1200 satoshis in fees.
+* Package {E, F, G, H, J} that pays 4000, 8000, 0, 2000, and 4000 satoshis in fees, respectively.
```
```suggestion
* Package {E, F, G, H, J} that pay 4000, 8000, 0, 2000, and 4000 satoshis in fees, respectively.
```

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+For example, consider the following scenarios (where the mempool minimum feerate is 3sat/vB and all
+transactions are the same size):
+
+* Package {A, B} where A pays 0 satoshis and B pays 8000 satoshis in fees.
```
Why not make the transactions' virtual size 1kvB to make the math super easy?

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+* Package {C, D} where C pays 0 satoshis and D pays 1200 satoshis in fees.
+* Package {E, F, G, H, J} that pays 4000, 8000, 0, 2000, and 4000 satoshis in fees, respectively.
+[[File:./bip-0331/tx_scenarios.png|800px]]
+
```
```suggestion
< br/>
```

The line below ("No Package Information Round...") wraps strangely on certain resolution displays. The "No Package" appears to the right of the image and then there's a new line for "Information Round...". I think you can force a line break so that the text appears on a new line.

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+child's wtxid to refer to the package and always send the entire package together. This would make
+it very likely for honest nodes to redownload duplicate transactions.
+[[File:./bip-0331/no_package_info.png|900px]]
+
```
As above, force a newline here:

```suggestion
<br />
```

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+
+====MSG_PKGTXNS====
+
+# A new inv type (MSG_PKGTXNS == 0x7) is added, for use only in "notfound" messages pertaining to
+package transactions.
```
```suggestion
# A new inv type (MSG_PKGTXNS == 0x7) is added, for use only in "notfound" messages pertaining to package transactions.
```

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+
+A "combined hash" serves as a unique "package id" for some list of unique transactions.
+
+The combined hash of a package of transactions is equal to the hash of each transaction's wtxid, concatenated in lexicographical order.
```
Which hash function?

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,391 @@
+
+# A node MUST NOT send a "ancpkginfo" message that has not been requested by the recipient. Upon receipt of an unsolicited "ancpkginfo", a node may disconnect the sender.
+
+# A "ancpkginfo" message should only be sent if both peers agreed to send version 0 packages. If a "ancpkginfo" message is received from a peer with which ancestor package relay was not negotiated, the sender may be disconnected.
```
It'd be good to explicitly define what a "version 0 package" is in this doc.

glozow commented:
@jnewbery thanks, took all suggestions

kallewoof commented:
Sorry about that, the BIP-327 title issue has been fixed, so please rebase.

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+batches.</ref>
+
+
+====ancpkginfo====
```
Would it make sense to put this above the definition of "getpkgtxns"? "ancpkginfo comes first both sequentially in the protocol flow and alphabetically.

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+# As a getdata request type, it indicates that the sender wants a "ancpkginfo" containing all of the unconfirmed ancestors of a transaction, referenced by wtxid.
+
+# Upon receipt of a "getdata(MSG_ANCPKGINFO)" request for, the node should respond with the "ancpkginfo" corresponding to the transaction's unconfirmed ancestor package, or with NOTFOUND.  The wtxid of the requested transaction should be the last item in the list. The node should not assume that the sender is requesting the transaction data as well.
```
```suggestion
# Upon receipt of a "getdata(MSG_ANCPKGINFO)" request, the node should respond with the "ancpkginfo" corresponding to the transaction's unconfirmed ancestor package, or with NOTFOUND.  The wtxid of the requested transaction should be the last item in the list. The node should not assume that the sender is requesting the transaction data as well.
```

jnewbery replied:
```suggestion
# Upon receipt of a "getdata(MSG_ANCPKGINFO)" request for, the node should respond with an "ancpkginfo" message corresponding to the transaction's unconfirmed ancestor package, or with NOTFOUND.  The wtxid of the requested transaction should be the last item in the list. The node should not assume that the sender is requesting the transaction data as well.
```

jnewbery replied:
I think "The node should not assume that the sender is requesting the transaction data as well." is unnecessary, since the description of the protocol above has already specified that the protocol has separate package information and transaction data rounds.

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+# Upon receipt of a "getdata(MSG_ANCPKGINFO)" request for, the node should respond with the "ancpkginfo" corresponding to the transaction's unconfirmed ancestor package, or with NOTFOUND.  The wtxid of the requested transaction should be the last item in the list. The node should not assume that the sender is requesting the transaction data as well.
+
+# The inv type should not be used in announcements, i.e. "inv(MSG_ANCPKGINFO)" should never be sent.
```
What should happen if a node receives an `inv(MSG_ANCPKGINFO)`? Drop/ignore the message or disconnect/punish?

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+# Upon receipt of a "ancpkginfo" message, the node may use it to request the transactions not already in its mempool.
+
+# Upon receipt of a malformed "ancpkginfo" message that does not contain the ancestor package of the transaction (which the node learns after receiving the transaction data and validating them), the sender may be disconnected.
```
This seems to slightly contradict the "Should a peer be punished if they provide incorrect package info" footnote, or am I misunderstanding?

glozow replied:
Fixed, described what malformed means

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+# The inv type should not be used in announcements, i.e. "inv(MSG_ANCPKGINFO)" should never be sent.
+
+# This inv type should only be used if both peers agreed to send version 0 packages. If a "getdata" message with type MSG_ANCPKGINFO is received from a peer with which ancestor package relay was not negotiated, the sender may be disconnected.
```
Perhaps:

```suggestion
# This inv type should only be used if both peers agreed to send version 0 packages. If a "getdata" message with type MSG_ANCPKGINFO is received from a peer with which ancestor package relay was not negotiated, no response should be sent and the sender may be disconnected.
```

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+# A node MUST NOT send a "ancpkginfo" message that has not been requested by the recipient. Upon receipt of an unsolicited "ancpkginfo", a node may disconnect the sender.
+
+# A "ancpkginfo" message should only be sent if both peers agreed to send version 0 packages. If a "ancpkginfo" message is received from a peer with which ancestor package relay was not negotiated, the sender may be disconnected.
```
This seems slightly redundant with the point above. If package relay has not been negotiated, then the peer shouldn't have sent a `getdata(MSG_ANCPKGINFO)`

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+# A "getpkgtxns" message should be used to request all or some of the transactions previously announced in a "ancpkginfo" message, specified by witness transaction id. This message is intended to allow nodes to avoid downloading and storing transactions that cannot be validated immediately.
+
+# Upon receipt of a "getpkgtxns" message, a node must respond with either a "pkgtxns" containing the requested transactions or "notfound" messages indicating one or more of the transactions is unavailable.<ref>'''Why isn't package relay negotiation required for getpkgtxns?'''
```
If I read this bullet point and footnote literally, it says that *all* nodes *must* respond to `getpkgtxns` messages, _even if they haven't negotiated package relay_. That seems wrong. It means that all existing nodes on the network would be in violation of this rule.

glozow replied:
Yes my bad, a node *should* respond.

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+well-researched sender-initiated package relay protocol.</ref>
+
+Ancestor package relay is negotiated between two peers during the version handshake using a
+"sendpackages" message containing version=1. It requires both peers to support wtxid-based relay
```
Should this be version=0?

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+end. See [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-May/020510.html discussion].
+</ref>
+
+# Version=0 packages correspond to ancestor packages, i.e. "ancpkginfo" and "MSG_ANCPKGINFO".
```
I think it'd be better to pull this definition out and make it more explicit. It'd also be useful to distinguish between the parts of this proposal that are specific to version 0 (`MSG_ANCPKGINFO` and `ancpkginfo`) and the parts that are generic to any kind of package (`sendpackages`, `getpkgtxns`, `pkgtxns` and `MSG_PKGTXNS`)

glozow replied:
Done

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+<ref>'''Will package relay cause non-package relay nodes to waste bandwidth on low-feerate transactions?'''
+If a node supports package relay, it may accept low-feerate transactions (e.g. paying zero fees)
+into its mempool, but non-package relay nodes would most likely reject them. To mitigate bandwidth
+waste, a package relay node should probably not announce descendants of below-fee-filter
```
Can we remove "probably" here?

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+==Implementation==
+
+Implementation for Bitcoin Core: https://github.com/glozow/bitcoin/tree/2023-01-package-relay
+
+Caveat: the crux of package relay is a mempool policy that works.
```
Caveat/crux/"that works" all feel a bit informal/imprecise. Perhaps:

```suggestion
Sample implementation for Bitcoin Core: https://github.com/glozow/bitcoin/tree/2023-01-package-relay. A prerequisite for implementing a safe package relay protocol is to have a mempool acceptance policy that is able to safely validate packages of transactions.
```

stickies-v commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+===Combined Hash===
+
+A "combined hash" serves as a unique "package id" for some list of unique transactions.
```
I think one use case to unambiguously determine if we've already validated (/rejected) a certain (exact) package as per [this comment](https://gist.github.com/glozow/8469dc9c3a003c7046033a92dd504329#:~:text=For%20packages%20and%20sub%2Dpackages%20that%20fail%2C%20add%20the%20transaction%20group%20by%20hashing%20the%20wtxids%20of%20each%20of%20the%20transactions%2C%20sorted%20lexicographically.%20Ensure%20that%20each%20group%20is%20itself%20an%20ancestor%20package.%20When%20a%20package%20is%20partially%20submitted%2C%20exclude%20the%20transactions%20that%20ended%20up%20in%20the%20mempool.)

glozow replied:
Yes, the combined hash is only used in `notfound(MSG_PKGTXNS)`.

We cannot use the child wtxid because the child wtxid is not included in the `getpkgtxns` request.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+A '''package''' is a list of transactions, representable by a connected Directed Acyclic
+Graph (a directed edge exists between a transaction that spends the output of another transaction).
+Typically, a package refers to unconfirmed transactions.
```
> What happens if my node sends a package including a confirmed ancestor transaction?

The receiver should be able to deal with it. It is not a violation to do so, as you might just have different chainstates.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+batches.</ref>
+
+
+====ancpkginfo====
```
> Do we want to redefine a package info message for every package version? 

Yes

> Could also use a generic pkginfo with version field matching the one in sendpackages?

I think it makes more sense to have a separate protocol message for separate things than to add 4-byte overhead on *every* package info message.

glozow commented:
Thanks @jnewbery @stickies-v @willcl-ark for the feedback. I'm working on incorporating the suggestions, apologies for the delay.

jnewbery commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+{|
+|  Field Name  ||  Type  ||  Size  ||  Purpose
+|-
+|version || uint32_t || 4 || Denotes a package version supported by the node.
```
I agree that having space for 4294967295 package versions is probably more than we'll ever need.

An alternative would be to treat this as a bit field, in the way that `nServices` flags are signalled in the version message. This BIP would define version `(1 << 0)` and future versions would be defined as `(1 << n)`. When we receive a `sendpackages` message, we XOR the received versions with the received versions from previous `sendpackages` messages.

glozow replied:
Yes I think a bitfield is best :+1: 32 package versions seems plenty, and this makes the implementation more cleaner as well.

glozow commented:
Changed the `version` field in "sendpackages" to `versions`, now interpreted as a bitfield, and updated the negotiation procedure accordingly. (https://github.com/bitcoin/bips/pull/1382#discussion_r1173605835)
Swapped the inv values for `MSG_ANCPKGINFO` and `MSG_PKGTXNS` (https://github.com/bitcoin/bips/pull/1382#discussion_r1156267093)

Neither of these are implemented in the branch I had linked before, so I've removed that link. Will update when I've done that.

I've restructured the BIP to make the difference between "generic" (can be used in future versions of package relay) vs "ancestor info" (first defined version) parts more clear.Also updated the diagrams and took wording suggestions.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+# The "sendpackages" message MUST be sent before sending a "verack" message. If a "sendpackages" message is received afer "verack", the sender may be disconnected.
+
+# A node that sends "sendpackages" MUST also send "wtxidrelay". Upon receipt of a "verack", if the sender has sent a "sendpackages" but not "wtxidrelay", the sender may be disconnected.
+
```
Removed

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+
+====MSG_PKGTXNS====
+
+# A new inv type (MSG_PKGTXNS == 0x7) is added, for use only in "notfound" messages pertaining to package transactions.
```
Done, I like it better too :)

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+information can be used to request transaction data. As these transactions are dependent upon one
+another to be valid, the transactions can be requested and sent as a batch.
+
+[[File:./bip-0331/protocol_flow.png|1200px]]
```
I omit the `MSG` because all the inv types have it

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,394 @@
+* Package {A, B} where A pays 0 satoshis and B pays 8000 satoshis in fees.
+* Package {C, D} where C pays 0 satoshis and D pays 1200 satoshis in fees.
+* Package {E, F, G, H, J} that pay 4000, 8000, 0, 2000, and 4000 satoshis in fees, respectively.
+[[File:./bip-0331/tx_scenarios.png|800px]]
```
Ended up removing this as I don't think it was adding very much

fjahr commented:
Maybe it's obvious or I am missing something but I couldn't see the answer in the doc: How would a node decide between sending `Inv(PKGINFO, C)` or `Inv(WTX, C)` in the Generic Package Relay example graphic? Is it always sending `PKGINFO` if the peer wants packages and C is part of a package?

instagibbs commented:
@fjahr "The inv type should not be used in announcements" it's a type used for `getdata` only IIUC

glozow commented:
Yes, `PKGINFO` is not specified anywhere; think of it as an "abstract class." The purpose of that diagram is to show how this protocol might be extended or upgraded in the future, i.e. we can define a new type of `PKGINFO` message to be used in invs and/or getdatas, and continue using the same getpkgtxns and pkgtxns message for transaction download. I want to separate those 2 concepts: exchanging package information and exchanging transaction data.

For example, `CHUNKPKGINFO` could contain a chunk (with cluster-based mempool), perhaps specifying the chunk fee and size as well. Originally, I proposed a `PCKGINFO1` type which was a child and its parents to communicate a fee-bumping relationship.

The reason there is an "inv(PKGINFO)" there is to communicate that, while `inv(ANCPKGINFO)` is not allowed, some other kind of `inv(PKGINFO)` can be allowed in the future for a sender-initiated protocol.

fjahr commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+ancestors.  Nodes indicate support for Ancestor Package Relay by adding the
+<code>PKG_RELAY_ANCPKG = (1 << 0)</code> bit in their "sendpackages" messages during version handshake.
+
+===Intended Protocol Flow===
```
While this section does talk about protocol flow it does so in the context of specific examples. I would probably call this section "Practical application of the Protocol with examples" or something like that, and move it out of and right after the specification section. I.e. have it between Specification and Compatibility. But I don't feel too strongly about this.

glozow replied:
I've changed this to "Protocol Flow Examples" but kept them in the same place. I'm thinking they might be helpful earlier on as a quick way to learn how the dialogue looks. Lmk if you disagree though!

fjahr reviewed (commented):
In my opinion, the proposed protocol is sound and the BIP provides almost all the necessary information. I am asking for a little more clarification in some places but most of my suggestions are intended to make to document easier to parse for first-time readers. While the information is there it is often ordered differently than I would expect. Maybe others' brains work differently but for me, this would have made a significant difference when I started studying this.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+can observe its usage and then introduce a sender-initiated package relay protocol informed by data
+collected from the p2p network.</ref>
+
+===Combined Hash===
```
Unsure... somebody would need to this in order to properly implement MSG_PKGTXNS. That seems like it should be "specification" ?

glozow replied:
Kept this as is for now

instagibbs commented:
bip-0331/package_cpfp_flow.png <--- looks like receiver is asking for A even though they already have it

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# Upon receipt of a "pkgtxns" message, the node should validate the transactions together as a package rather than individually.
+
+# A "pkgtxns" message should only be sent to a peer that requested the package using "getpkgtxns".  If a node receives an unsolicited package, the sender may be disconnected.<ref>'''Why isn't package relay negotiation required for pkgtxns?'''
```
Yes, reason to disconnect. I think that is the same thing as an unsolicited package.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+In a '''topologically sorted''' package, each parent appears somewhere in the list before its child.
+
+==Specification==
+
```
Thanks, yeah, it's weird the way drew a line between generic/ancestor package relay!

Had an offline conversation with someone about the "generic" thing. I decided to get rid of that and instead add the information in an "Extensibility" section. At the end of the day, I'm just suggesting what a future package relay upgrade could look like and we are not locked into using the format I outlined.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+continuing to use the same messages for transaction data download.
+[[File:./bip-0331/generic_package_relay.png|600px]]
+
+Package relay is negotiated between two peers during the version handshake using a "sendpackages"
```
:+1: I've gotten rid of Generic and hopefully the flow is now more natural since there's only 1 section in "New Messages"

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+===Ancestor Package Information===
+
+Ancestor Package Relay allows nodes to send and request a transaction's list of unconfirmed
```
Changed all the proper noun instances to Ancestor Package Relay

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# The "sendpackages" message MUST be sent before sending a "verack" message. If a "sendpackages" message is received afer "verack", the sender may be disconnected.
+
+# Upon successful connection ("verack" sent by both peers), a node may relay packages with the peer if they did not set "fRelay" to false in the "version" message, both peers sent "wtxidrelay", and both peers sent "sendpackages" for matching version(s). Peers should relay packages corresponding to versions that both sent "sendpackages" for.<ref>'''Is it ok to send "sendpackages" to a peer that sp ...
```
Yes, though ignored. I've added a note about this now

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# The "sendpackages" message MUST be sent before sending a "verack" message. If a "sendpackages" message is received afer "verack", the sender may be disconnected.
+
+# Upon successful connection ("verack" sent by both peers), a node may relay packages with the peer if they did not set "fRelay" to false in the "version" message, both peers sent "wtxidrelay", and both peers sent "sendpackages" for matching version(s). Peers should relay packages corresponding to versions that both sent "sendpackages" for.<ref>'''Is it ok to send "sendpackages" to a peer that sp ...
```
Thanks, taken

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# The "sendpackages" message has the structure defined above, with pchCommand == "sendpackages".
+
+# During version handshake, nodes should one "sendpackages" message indicating they support package relay, with the versions field indicating which versions of they support.
```
Thanks, fixed

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+[[File:./bip-0331/version_negotiation.png|400px]]
+
+===Ancestor Package Information===
```
Heh, I think in the future if we have 2, we could probably say `PKG_RELAY_ANC` (the name of the version bit) to refer to this one?

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# A "getpkgtxns" message should be used to request some list of transactions specified by witness transaction id. It indicates that the node wants to receiver either all the specified transactions or none of them. This messages is intended to allow nodes to avoid downloading and storing transactions that cannot be validated without each other. The list of transactions does not need to correspond  ...
+
+# Upon receipt of a "getpkgtxns" message, a node should respond with either a "pkgtxns" containing all of the requested transactions in the same order specified in the "getpkgtxns" request or one "notfound" message of type MSG_PKGTXNS and combined hash indicating one or more of the transactions is unavailable.<ref>'''Why isn't package relay negotiation required for getpkgtxns?'''
```
Yep! tried to clarify this

glozow commented:
> bip-0331/package_cpfp_flow.png <--- looks like receiver is asking for A even though they already have it

Thanks for catching!!! Fixed

fjahr commented:
ACK, thanks for the updates, I find it an easier read now and didn't see any other issues in my latest pass.

One more question: this protocol seems pretty convenient for fingerprinting nodes by giving specific ancestors to different connections and then observing which tx the connections request to complete the package. I don't think this is a huge problem since there are several other options that allow fingerprinting. But I am curious if this has been discussed somewhere in the past or if there are already ideas for mitigations in the implementation (which I haven't fully reviewed yet).

instagibbs commented:
@fjahr can you think of a fundamentally new vectors vs the back and forth orphan fetching? Package fetching seems to be more of a optimization on that. Higher resolution fingerprinting, if it's more reliable I guess :)

glozow commented:
@fjahr fingerprinting is a good consideration but neither new nor made
easier. An even more convenient approach: make invalid transactions (as
many as you want for free), each with different wtxids, and send one to
each node. Nodes will cache their unique rejections and not send a getdata
on announcement.

El El lun, 5 jun 2023 a las 21:47, Fabian Jahr ***@***.***>
escribió:

> @fjahr <https://github.com/fjahr> can you think of a fundamentally new
> vectors vs the back and forth orphan fetching? Package fetching seems to be
> more of a optimization on that. Higher resolution fingerprinting, if it's
> more reliable I guess :)
>
> I don't think it's fundamentally new, that's why I don't think it's a big
> issue. But I do think it makes the attack more convenient to scale.
>
> What I had in mind: 1. Make an unconfirmed tx chain A through G. 2. Give
> different connections different combinations of the txs B through F. This
> should be 5! = 120 unique fingerprints. 3. Send G and respond to the getdata(ANCPKGINFO,
> G) to see which tx of B through F the connection is aware of.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bips/pull/1382#issuecomment-1577450727>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGAEGGKBQPRGH6NQJYKJNG3XJZAVLANCNFSM6AAAAAAREVY6QA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,426 @@
+probably need to request and validate a group of transactions together in order for them to be
+accepted. As with any proactive communication, there is a chance that the receiver already knows
+this information, so this network bandwidth may be wasted. Shortened latency is less significant
+than wasted bandwidth.
```
I think this assumption is dependent on the bumping patterns of package broadcast. In the sense that if the package is re-broadcast as it is (and there is no mempool fluctuations), there is a complete waste of bandwidth as both parent and child are known.

If the package is rebroadcast with the CPFP bumped (probably the future usage for LN where the parent commitment transaction stays an invariant), there is only a need to re-announce the child to the peer nodes.

And note bandwidth assumptions we’re making on our side can be falsified in changes in the re-broadcast logic of LN implementations, like it has been done recently.

I think a realistic simulation model would encompass at least few dimensions 
- the peer-to-peer announcement format
- the mempool fluctuations and evictions of package component
- the rebroadcast frequency of applications (and if there is a malleation of CFPP at each rebroadcast)

More dimensions to think about ?

glozow replied:
Yeah so sender-initiated is not being proposed here. The general philosophy is that since everything here is receiver-initiated in certain situations, and we are careful to not waste bandwidth in honest cases in the implementation, hopefully we are not blowing up bandwidth.

If we can do even better by doing sender-initiated in some cases in the future, that'd be nice. I don't think any simulation can be realistic, as we have 0 data about package relay usage until we actually deploy it and people use it to relay transactions that they couldn't before. But these are very helpful considerations for measuring things, thanks will keep in mind.

ariard reviewed (commented):
Review at “Combined Hash” excluded.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,426 @@
+<br />
+
+'''No Package Information Round:''' One proposal is to just use the child's wtxid to refer to the
+package and always send the entire package together, skipping the package information round.
```
Note that this comment is on an alternative design considered but *not* proposed.

> And I think such design would suffer from transaction censorship vector if you have a parent with multiple childs of unequal feerate ?

> you can attach your to_remote child downgrading the feerate of the whole package.

Sure, an additional child may exist. But it is not included in the package because it's not part of the actual cpfp child's ancestor set. Its low fees should not be considered in combination with transactions that don't rely on it.

As for potential censorship, that is solved by caching "reconsiderable" rejections separately (see `m_recent_rejects_reconsiderable` in PR). You can relay a separate package that is the ancestor set of the additional child, and its rejection should not affect whether another overlapping ancestor package is considered/accepted.

ariard replied:
> As for potential censorship, that is solved by caching "reconsiderable" rejections separately (see m_recent_rejects_reconsiderable in PR). You can relay a separate package that is the ancestor set of the additional child, and its rejection should not affect whether another overlapping ancestor package is considered/accepted.

Yes sounds correct to process each cpfp’s ancestor set separately.

instagibbs commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,426 @@
+"sendpackages".  A future version of package relay may allow a sender-initiated dialogue if it is
+specified that the "*PKGINFO" inv type can be used in an "inv" message.
+<br />
+[[File:./bip-0331/generic_package_relay.png|600px]]
```
Actually not totally clear what this image is showing. why is there a notfound followed by packagetxns?

glozow replied:
Changed it and added some more descriptions. I had wanted to show multiple possible dialogues but it just looks confusing.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,426 @@
+<ref>'''Why do we need multiple versions? Why can't we just support arbitrary packages?'''
+Attempting to support arbitrary packages in mempool validation may result in very complex logic, new
+Denial of Service attack vectors, and policy limitations that could be leveraged to censor
+transactions (aka "pinning attacks"). This protocol is extensible to support up to 32 types of
```
> It sounds that type of cheap bandwidth saving that might reveal as a downside a decade from now ? And it’s one time message per-link.

> Let’s hypothetize you will have half a dozen of time-sensitive second-layers deployed few years from now (lightning, vaults, statechains, submarine swaps, time-locked wallets, discreet log contracts). Among those second-layers, they might need different variants of packages for each of their protocol versions (e.g for Lightning taproot channels, eltoo channels, demux channels for jamming) and for each protocol phase (dual-funding, splicing).

Er, what exactly is the "cheap bandwidth saving?" Are you suggesting to increase the field size?

I've removed the mention of 32 versions to leave room for other ways to extend. For now, I don't think we should over-engineer the versions field for a hypothetical super-complex future that may or may not need dozens of types of packages. I think 32 is plenty, and the worst case scenario is defining a new message and sending another few bytes once per connection.

ariard replied:
> Er, what exactly is the "cheap bandwidth saving?" Are you suggesting to increase the field size?

Yes increasing the filed size to a u64 or u128 to have more bits available for mechanism.

>  I think 32 is plenty, and the worst case scenario is defining a new message and sending another few bytes once per connection.

Sure we can always define a new message, or an extension of it at the price of more p2p code surface and complexity (and note all the bikeshedding with `fRelay`). Sounds package relay will get more complex not less, though happy to take the bet and wait when this hypothetical super-complex future does happen :)

ajtowns commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,428 @@
+<ref>'''Is Package Erlay possible?'''
+A client using BIP330 reconciliation-based transaction relay (Erlay) is able to use package relay
+without interference. After reconciliation, any transactions with unconfirmed ancestors may be
+relayed using ancestor package relay.
```
We already have the transaction at that point, so maybe better to phrase it as "any transaction with unconfirmed ancestors may have those ancestors resolved using ancestor package relay" ?

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,428 @@
+Ancestor Package Relay includes two parts: package information and transaction data download.
+A package information round is used to help a receiver learn what transactions are in a package and
+whether they want to download them. The transaction data round is used to help a node download
+multiple transactions in one message instead of as separate messages.
```
I don't think it's that simple - we have cases of
- I downloaded this tx and rejected it for too low feerate, but I'm okay with trying again because it's with a package of other transactions I haven't seen before
- I haven't seen this tx before but there's another tx in the package that I know is consensus-invalid so I'm not going to download any of them

ajtowns replied:
> In future package types where you might get additional information, e.g., package feerates or similar this may make more sense to say.

Okay, "the information round *described here* doesn't really help you learn..."

> I haven't seen this tx before but there's another tx in the package that I know is consensus-invalid so I'm not going to download any of them

In that case, (on the implementation side) would we need different reject caches, one for "consensus invalid", one for "policy invalid"? Not sure that would be worth the hassle (honest peers sending consensus invalid stuff should be extremely rare; attackers could just send novel txs each time avoiding the cache)

glozow replied:
> In that case, (on the implementation side) would we need different reject caches, one for "consensus invalid", one for "policy invalid"?

Yes. Two rejection caches is necessary, see https://github.com/glozow/bitcoin/pull/8#issuecomment-1453791776. But not consensus vs policy, just ones that are eligible for reconsideration in a package (i.e. too low feerate or missing ephemeral anchor spend) vs ones that are not. I said "consensus invalid" to provide a clear example, but I just mean "something we don't need to validate again, even if it's in a package." If something was larger than max standard size or has too many ancestors, trying it in a package won't make it valid.

ajtowns replied:
I guess as long as they're global caches and not per-peer it doesn't matter much anyway.

glozow replied:
I'm going to resolve this for now, but let me know if you still feel that I should change this. Right now the text is "The package information round is used to help a receiver learn what transactions are in a package and decide whether they want to download them."

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,428 @@
+
+# A "getpkgtxns" message should be used to request some list of transactions specified by witness transaction id. It indicates that the node wants to receive either all the specified transactions or none of them. This message is intended to allow nodes to avoid downloading and storing transactions that cannot be validated without each other. The list of transactions does not need to correspond to ...
+
+# Upon receipt of a "getpkgtxns" message, a node should respond with either a "pkgtxns" containing all of the requested transactions in the same order specified in the "getpkgtxns" request or one "notfound" message of type MSG_PKGTXNS and combined hash of all of the wtxids in the "getpkgtxns" request (only one "notfound" message and nothing else), indicating one or more of the transactions is una ...
```
> What if you received a getpkgtxns for wtxids that weren't in an ancpkginfo that you set?

That's fine imo - we'll also respond to getdata(tx) for transactions we didn't announce under certain conditions.

> Can you say "I'll never accept a tx with more than 20 ancestors in my mempool, so I'll never send more than 21 wtxids in an ancpkginfo, so I'll never send more than 21 txs in response to a getpkgtxns?

I think it's better for ancpkginfo and getpkgtxns logic to be independent, as the idea is to be able to reuse getpkgtxns for other package info types in the future. One might imagine a future where we have cluster package info (or something), and even though ancestor limit is 21, there is a cluster with 22 transactions.

ajtowns replied:
I agree in principle, but I think there probably should be some explicit limit for what you can put in getpkgtxns. With 65 byte txs being standard, you could fit ~60k txs in a pkgtxns response; is there anything beyond "whoops, your response would exceed `MAX_PROTOCOL_MESSAGE_LENGTH` so you get a notfound instead" that we could use to trim nonsense requests early? There's a big difference between 21 and 60k!

glozow replied:
Ah good point. Perhaps we limit to, say, 50? 100? (Arbitrarily picked, 25 just seems a little bit too limiting)

glozow replied:
Marking as resolved based on https://github.com/bitcoin/bips/pull/1382#discussion_r1235492806

ajtowns commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,428 @@
+
+# Upon receipt of a "pkgtxns" message, the node should validate the transactions together as a package rather than individually.
+
+# A "pkgtxns" message should only be sent to a peer that requested the package using "getpkgtxns".  If a node receives an unsolicited package, the sender may be disconnected.<ref>'''Why isn't package relay negotiation required for pkgtxns?'''
```
The way I would read the above is that most of those "mays" would be violating the previous paragraph: "the node should validate the transactions together as a package rather than individually".

glozow replied:
Do I understand correctly that you're saying we might want to validate an unsolicited package and not disconnect the sender? Why would we do this? It doesn't seem like a good idea to spend resources validating transactions if we didn't ask for them.

instagibbs replied:
> "the node should validate the transactions together as a package rather than individually".

hm good point. I'm not actually sure what this phrase means? Even the implementation for Core tries transactions individually and in subpackages.

> Why would we do this? It doesn't seem like a good idea to spend resources validating transactions if we didn't ask for them.

It's kind of an implementation detail, in that it doesn't effect the underlying protocol(?)

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,428 @@
+Package relay is negotiated between two peers during the version handshake using a "sendpackages"
+message. The versions field within "sendpackages" is interpreted as a bitfield; peers may relay
+multiple versions of packages. Package relay requires both peers to support wtxid-based relay
+because package transactions are referenced by their wtxids.
```
Ah good idea.

instagibbs commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,428 @@
+
+# A "pkgtxns" message should contain the transaction data requested using "getpkgtxns".
+
+# Upon receipt of a "pkgtxns" message, the node should validate the transactions together as a package rather than individually.
```
https://github.com/bitcoin/bips/pull/1382#discussion_r1229640816

As per discussion, I'm not sure what this sentence actually means. It's up to the node implementation to decide how to handle these, including trying individual validation first, then subpackage subsets...

glozow replied:
Deleted

glozow commented:
Updated for suggestions from @ajtowns, thanks!

Major changes (still need to update implementation to reflect these):
- Added a `PKG_RELAY_PKGTXNS = 1 << 0` and `PKG_RELAY_ANC` is now `1 << 1` (signaling both is versions=3). (https://github.com/bitcoin/bips/pull/1382#discussion_r1225156027)
- Set maximum 100 wtxids in a `getpkgtxns` request. (https://github.com/bitcoin/bips/pull/1382#discussion_r1228918874)

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# Upon receipt of a "getpkgtxns" message, a node should respond with either a "pkgtxns" containing all of the requested transactions in the same order specified in the "getpkgtxns" request or one "notfound" message of type MSG_PKGTXNS and combined hash of all of the wtxids in the "getpkgtxns" request (only one "notfound" message and nothing else), indicating one or more of the transactions is una ...
+
+# A "getpkgtxns" message must contain at most 100 wtxids. Upon receipt of a "getpkgtxns" message with more than 100 wtxids, a node may ignore the message (to avoid calculating the combined hash) and disconnect the sender.
```
No worries. I picked 100 arbitrarily, but I don't imagine we'd be picking a cluster limit higher than that (just a guess, maybe somebody else has more insight). Note that this max is also the number of wtxids we may potentially need to sort + hash together to send a "notfound" response.

ajtowns replied:
I'd be pretty cautious about trying to request large chunks via getpkgtxns -- if any tx has been bumped, you'll just get a notfound with no indication of what part might be okay to retry.

instagibbs replied:
sounds like clusters are unlikely to go this high anyways, so let's keep it at 100 for this and the other reasons above

instagibbs replied:
https://github.com/bitcoin/bitcoin/pull/28676/commits/4c9842999ed5e35d959f10185a7ca82194ec28c5#diff-1fc0f6b5081e8ed5dfa8bf230744ad08cc6f4c1147e98552f1f424b0492fe9bdR59 for reference to initial cluster mempool implementation

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# The "ancpkginfo" message has the structure defined above, with pchCommand == "ancpkginfo".
+
+# The "txns" field should contain a list of wtxids which constitute the ancestor package of the last wtxid. The sender should - but is not required to - sort these wtxids in topological order. The topological sort can be achieved by sorting the transactions by mempool acceptance order (if parents are always accepted before children). Nodes should not disconnect or punish a peer who provides a lis ...
```
Side-note, I think the feerate cannot be trusted under BIP118  [ANYPREVOUTANYSCRIPT,](https://github.com/bitcoin/bips/blob/master/bip-0118.mediawiki) where the spent amount are not committed in the digest and transaction child A can be attached to parent X or parent Z of different amounts (of course there is the issue to re-bind the ANYPREVOUTANYSCRIPT on a parent by scriptpubkey only…). So I think this avoid to introduce a dependency in light of potential future consensus changes.

glozow replied:
Useful extra context, though I didn't add any of this to the text.

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+|-
+|txns_length||CompactSize||1 or 3 bytes|| The number of transactions provided.
+|-
+|txns||List of wtxids||txns_length * 32|| The wtxids of each transaction in the package.
```
There is no limit on the number of announced ancestors ?

From my understanding, there is a bandwidth asymmetry in this design. A receiver can request an unbounded number of parents from the sender by sending a single GETDATA(wtixd), where the child transaction is deliberately spending a large number of in-mempool unconfirmed ancestors. As the ancestors are not linked without the child, I think the number can be beyond in-mempool ancestors limits.

Those ancestors might have been previously announced by an inbound connection controlled by the adversary for a bandwidth cost of O(n), however the target node might have to spend a bandwidth cost of O(n*m), where m is the number of transaction-relay connections controlled by the adversary, if I’m correct ?

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# Upon receipt of a malformed "ancpkginfo" message, the sender may be disconnected. An "ancpkginfo" message is malformed if it contains duplicate wtxids or conflicting transactions (spending the same inputs). The receiver may learn that a package info was malformed after downloading the transactions.
+
+# A node MUST NOT send a "ancpkginfo" message that has not been requested by the recipient. Upon receipt of an unsolicited "ancpkginfo", a node may disconnect the sender.
```
If the wtxid of the requested transaction is missing in the list, and therefore the “ancpkginfo” is uncomplete, should the peer be disconnected (as there is no way to dissociate it’s a unsolicited package) ?

glozow replied:
Yes, that is an unsolicited ancpkginfo.

glozow replied:
Added an extra sentence about the fact that the last transaction is required to be the transaction we're talking about.

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# The "getpkgtxns" message has the structure defined above, with pchCommand == "getpkgtxns".
+
+# A "getpkgtxns" message should be used to request some list of transactions specified by witness transaction id. It indicates that the node wants to receive either all the specified transactions or none of them. This message is intended to allow nodes to avoid downloading and storing transactions that cannot be validated without each other. The list of transactions does not need to correspond to ...
```
“previsously-received “ancpkginfo” message” for more clarity i think

glozow replied:
Done

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# A "getpkgtxns" message should be used to request some list of transactions specified by witness transaction id. It indicates that the node wants to receive either all the specified transactions or none of them. This message is intended to allow nodes to avoid downloading and storing transactions that cannot be validated without each other. The list of transactions does not need to correspond to ...
+
+# Upon receipt of a "getpkgtxns" message, a node should respond with either a "pkgtxns" containing all of the requested transactions in the same order specified in the "getpkgtxns" request or one "notfound" message of type MSG_PKGTXNS and combined hash of all of the wtxids in the "getpkgtxns" request (only one "notfound" message and nothing else), indicating one or more of the transactions is una ...
```
Here I think you’re restraining some partial package download, where none of your peers have all the ancestors for a known child transaction however the unions of all the available transactions announced by our peers give you such complete package.

I think that’s okay and the behavior of “notfound(MSG_PKGTXND)” can be changed by bumping the version bits.

glozow replied:
> where none of your peers have all the ancestors for a known child transaction however the unions of all the available transactions announced by our peers give you such complete package.

We should be able to hope that at least one of the peers was able to keep the list of wtxids they told us about in their ancpkginfo message. If they evicted some ancestors, they also evicted the child we're trying to resolve with them.

Also, this definition of how the "getpkgtxns" message works does not preclude having logic for retrieving packages from multiple peers in your implementation.

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# A "pkgtxns" message should contain the transaction data requested using "getpkgtxns".
+
+# A "pkgtxns" message should only be sent to a peer that requested the package using "getpkgtxns". If a node receives an unsolicited package, it may choose to validate the transactions or not, and the sender may be disconnected.
```
I think here the receiver behavior can be more liberal and unsolicited “pkgtxns” shouldn’t be treated as a disconnection. A sender implementing some “prediction algorithm”, in the sense knowing the parent is going to be rejected on its own (e.g due to feerate under feefilter) might optimize by sending a full-package, without opt-in from your peer. This behavior could be legit as it’s RTT/bandwidth consumed on its side.

glozow replied:
I can see a "prediction algorithm" for proactive, sender-initiated *announcements* being appropriate and more effective. But I don't see why it would make sense to do this for full transaction data. Can you provide a specific reason why this would be needed?

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# As a notfound type, it indicates that the sender is unable to send all the transactions requested in a prior "getpkgtxns" message. The hash used is equal to the combined hash (defined above) of the wtxids in the getpkgtxns request.
+
+# This inv type should only be used in "notfound" messages, i.e. "inv(MSG_PKGTXNS)" and "getdata(MSG_PKGTXNS)" must never be sent.
```
Can precise peer will be disconnected.

glozow replied:
Added

ariard reviewed (commented):
See the note on maybe asymmetry around “ancpkginfo"

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,428 @@
+suggests adding new p2p messages enabling nodes to request and share package-validation-related
+information with one another, resulting in a more efficient and reliable way to propagate packages.
+
+(2) Eliminate the need for txid-based transaction relay and make orphan handling more robust.
```
I suppose it's true that this protocol is not going to have fewer round trips in every possible case of orphans, but we wouldn't have a way of detecting those cases on the receiver end.

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,428 @@
+
+# As a getdata request type, it indicates that the sender wants a "ancpkginfo" containing all of the unconfirmed ancestors of a transaction, referenced by wtxid.
+
+# Upon receipt of a "getdata(MSG_ANCPKGINFO)" request, the node should respond with an "ancpkginfo" message corresponding to the transaction's unconfirmed ancestor package, or with "notfound".  The wtxid of the requested transaction should be the last item in the list.
```
Updated to clarify that the last transaction must be the requested one, the rest can be in any order (nodes should not disconnect or punish), but sender "should" sort topologically for the receiver's convenience, and that can be achieved through sorting by acceptance order.

glozow commented:
I think I've addressed all comments now.

ariard commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# Upon receipt of a "getpkgtxns" message, a node should respond with either a "pkgtxns" containing all of the requested transactions in the same order specified in the "getpkgtxns" request or one "notfound" message of type MSG_PKGTXNS and combined hash of all of the wtxids in the "getpkgtxns" request (only one "notfound" message and nothing else), indicating one or more of the transactions is una ...
+
+# A "getpkgtxns" message must contain at most 100 wtxids. Upon receipt of a "getpkgtxns" message with more than 100 wtxids, a node may ignore the message (to avoid calculating the combined hash) and disconnect the sender.
```
Can precise `GETPKGTXNS` must not be empty, enforced as a disconnection condition (`Misbehaving(*peer, 100, …)`) on `GETPKGTXNS` reception (if i’m on the right commit).

glozow replied:
I don't think it's implemented this way? `if (num_txns == 0) return;` means it's ignored.

ariard replied:
Yeah would have to check back the code here.

instagibbs commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+
+# A new inv type (MSG_PKGTXNS == 0x6) is added, for use only in "notfound" messages pertaining to package transactions.
+
+# As a notfound type, it indicates that the sender is unable to send all the transactions requested in a prior "getpkgtxns" message. The hash used is equal to the combined hash (defined above) of the wtxids in the getpkgtxns request.
```
```suggestion
# As a notfound type, it indicates that the sender is unable to send all the transactions requested in a prior "getpkgtxns" message. The hash used is equal to the combined hash of the wtxids in the getpkgtxns request.
```

instagibbs commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+# The "ancpkginfo" message has the structure defined above, with pchCommand == "ancpkginfo".
+
+# The "txns" field should contain a list of wtxids which constitute the ancestor package of the last wtxid. For the receiver's convenience, the sender should - but is not required to - sort the wtxids in topological order. The topological sort can be achieved by sorting the transactions by mempool acceptance order (if parents are always accepted before children). Apart from the last wtxid which i ...
+Peers are not trusted to provide correct information, so a reported low feerate should not stop a
```
I'm not sure this first half is very persuasive, a node can lie and waste our bandwidth regardless. I think the latter half of the answer is the most sensible.

instagibbs commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,430 @@
+A "combined hash" serves as a unique "package id" for some list of transactions and helps provide a
+meaningful but short "notfound" response to "getpkgtxns."
+
+The combined hash of a package of transactions is equal to the double-sha256 hash of each transaction's
```
we could just make this single-sha2...

glozow replied:
Changed

instagibbs reviewed (approved):
ACK

nits only

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,426 @@
+* [https://github.com/lightning/bolts/blob/master/03-transactions.md#htlc-timeout-and-htlc-success-transactions HTCL-Timeout in LN Penalty]
+* [https://github.com/revault/practical-revault/blob/master/transactions.md#cancel_tx Unvault Cancel in Revault]
+* [https://github.com/discreetlogcontracts/dlcspecs/blob/master/Transactions.md#refund-transaction Refund Transaction in Discreet Log Contracts]
+* [https://gist.github.com/instagibbs/60264606e181451e977e439a49f69fe1 Updates in Eltoo]
```
Added

glozow reviewed (commented):
Addressed the text-related comments

glozow commented:
Changed combined hash to be single instead of double sha256

instagibbs commented:
re-ACK https://github.com/bitcoin/bips/pull/1382/commits/02ec218c7857ef60914e9a3d383b68caf987f70b

suggested textual additions and change to single-sha2 for combined hash

glozow commented:
@luke-jr @kallewoof is anything else needed here?

glozow commented:
@luke-jr pinging again, can we get this merged?

sr-gi reviewed (commented):
A couple of comments/nits I found when going over this

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,429 @@
+
+# A node MUST NOT send a "ancpkginfo" message that has not been requested by the recipient. Upon receipt of an unsolicited "ancpkginfo", a node may disconnect the sender.
+
+# This message must only be used if both peers both set <code>PKG_RELAY_ANC</code> in their "sendpackages" message. If an "ancpkginfo" message is received from a peer with which this type of package relay was not negotiated, no response should be sent and the sender may be disconnected.
```
Oh yes, this is a typo - I have an extra 'both'. Thanks!

sr-gi replied:
Note that this is not the only place in the doc where "Both peers both set" appears

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,429 @@
+
+# A "pkgtxns" message should only be sent to a peer that requested the package using "getpkgtxns". If a node receives an unsolicited package, it may choose to validate the transactions or not, and the sender may be disconnected.
+
+# This message must only be used if both peers both set <code>PKG_RELAY_PKGTXNS</code> in their "sendpackages" message. If a "getpkgtxns" message is received from a peer with which this type of package relay was not negotiated, no response should be sent and the sender may be disconnected.
```
I think "getpkgtxns" is correct actually?

sr-gi replied:
Is it? This is part of the "pkgtxns" section, so it would make sense to be referring to the message that is being talked about. That is, at least, how this note is handled in every other message section that includes it throughout the doc.

Also notice that, as is, this is the exact same text as in point 5 of "getpkgtxns"

glozow replied:
Ah we were both wrong. I've changed it to "pkgtxns", thanks

glozow commented on bip-0331.mediawiki:
```diff
@@ -0,0 +1,429 @@
+
+====MSG_PKGTXNS====
+
+# A new inv type (MSG_PKGTXNS == 0x6) is added, for use only in "notfound" messages pertaining to package transactions.
```
Added quotes for all notfounds. thanks!

glozow commented:
pinging again, if any BIP editors are available to have a look?
